### PR TITLE
Kratos rom save rom coeff process

### PR DESCRIPTION
**Pull request by SADPR**
## Introduction of SaveRomCoefficientsProcess

This PR introduces the `SaveRomCoefficientsProcess` for the ROM Application, designed to save the generalized coordinates (q) of the ROM to disk. It facilitates the recording of ROM state snapshots, offering both cumulative and incremental saving strategies based on user preference.

### Features

- **Snapshot Control**: Allows for time-based and step-based snapshot output, providing flexibility in the frequency of ROM state recording.
- **Output Customization**: Users can specify the output directory and file name for ROM coefficients, aiding in effective data management.
- **Saving Options**: Supports the choice between saving ROM coefficients cumulatively (`total`) or incrementally (`incremental`).

### Enhancements and Fixes

- **Test Addition**: Incorporated a new test that reuses the structural static case to validate the correct printing of the ROM coefficients (q). 
  
- **Parameter Handling Fix**: Addressed an assumption in `rom_analysis.py` where the `calculate_rom_basis_output_process` was expected to be the only ROM output process. Previously, parameters were directly accessed assuming a single process type, leading to potential errors when introducing new types of ROM output processes. The code now iterates through `rom_output_parameters` to dynamically identify and handle different processes, enhancing flexibility and fixing a critical limitation.

### CODE REVIEW

File: applications/RomApplication/tests/test_save_rom_coefficients_process.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            rom_coeffs_obtained_output_name = os.path.join(self.output_folder, "ObtainedOutputSaveRomCoefficientsProcess.npy")
+            rom_coeffs_obtained_output = np.load(rom_coeffs_obtained_output_name)
+
+            # Load reference file
+            rom_coeffs_expected_output_name = "../../save_rom_coefficients_process_test_files/ExpectedOutputSaveRomCoefficientsProcess.npy"
+            rom_coeffs_expected_output = np.load(rom_coeffs_expected_output_name)
+
+            for i in range(rom_coeffs_obtained_output.shape[0]):
+                for j in range(rom_coeffs_obtained_output.shape[1]):
+                    self.assertAlmostEqual(rom_coeffs_obtained_output[i,j], rom_coeffs_expected_output[i,j])
```
Comment (philbucher) [inline applications/RomApplication/tests/test_save_rom_coefficients_process.py]: `self.assertMatrixAlmostEqual`?

File: applications/RomApplication/python_scripts/save_rom_coefficients_process.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        self.model_part = model[settings["model_part_name"].GetString()]
+
+        # Set the snapshots output control and interval
+        snapshots_control_type = settings["snapshots_control_type"].GetString()
+        if snapshots_control_type == "time":
+            self.snapshots_control_is_time = True
+        elif snapshots_control_type == "step":
+            self.snapshots_control_is_time = False
+        else:
+            err_msg = "Unknown value \'{}\' for \'snapshots_control_type\'. Available options are \'time\' and \'step\'.".format(snapshots_control_type)
```
Comment (philbucher) [inline applications/RomApplication/python_scripts/save_rom_coefficients_process.py]:
```suggestion
            err_msg = f"Unknown value \'{snapshots_control_type}\' for \'snapshots_control_type\'. Available options are \'time\' and \'step\'."
```

File: applications/RomApplication/tests/test_save_rom_coefficients_process.py (hunk 1 above)
Comment (SADPR) [reply applications/RomApplication/tests/test_save_rom_coefficients_process.py]: Done.

File: applications/RomApplication/python_scripts/save_rom_coefficients_process.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            self.rom_snapshots.append(numpy.copy(self.cumulative_rom_state))
+        elif self.snapshot_solution_type=="incremental":
+            self.rom_snapshots.append(delta_q)
+
+        # Schedule the next output
+        current = self.model_part.ProcessInfo[KratosMultiphysics.TIME if self.snapshots_control_is_time else KratosMultiphysics.STEP]
+        while self.__GetPrettyFloat(self.next_output) <= self.__GetPrettyFloat(current):
+            self.next_output += self.snapshots_interval
+
+    def _PrintRomBasis(self):
```
Comment (Rbravo555) [inline applications/RomApplication/python_scripts/save_rom_coefficients_process.py]: Out of consistency, I'd call this method differently, e.g. PrintRomCoefficients()

**Review by Rbravo555 — COMMENTED** [review]
I am not sure the test is being called when launching python3 test_RomApplication.py

**Comment by SADPR** [comment]
@Rbravo555 you were right. The test was not properly called. Done.

**Review by Rbravo555 — APPROVED** [review]
LGTM

**Pull request merged**